### PR TITLE
UIMARCAUTH-29 Browse MARC Authority Records > Results List Sort behavior AND Pagination behavior

### PR DESCRIPTION
**Pull request by BogdanDenis**
## Description
Added Browse Authority records functionality

## Approach
Added `AuthoritiesSearchContext` that holds state about selected search options and value, filters, navigation segment, advanced search rows etc. Components that need access to this data/their setters can use the context to access what they requre.
Added `authoritiesCtxValue` to `<Harness>`. You can pass mocked context value to tested components by using this property on `<Harness>`
Added `useDidUpdate` hook that mimicks `componentDidUpdate` lifecycle method. Passed callback will not be called upon mounting unlike when simply using `useEffect`

## Screenshots

https://user-images.githubusercontent.com/19309423/154241221-bc5a8b1f-097d-4ec5-a608-d6d04c8101ab.mp4

## Issues
[UIMARCAUTH-29](https://issues.folio.org/browse/UIMARCAUTH-29)

### CODE REVIEW

File: src/components/AuthoritiesSearchForm/AuthoritiesSearchForm.js [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    setAdvancedSearchRows,
+    resetAll,
+  } = useContext(AuthoritiesSearchContext);
+
+  const [isAdvancedSearchOpen, setIsAdvancedSearchOpen] = useState(false);
+  const [advancedSearchDefaultSearch, setAdvancedSearchDefaultSearch] = useState({
+    query: locationSearchParams.query,
+    option: locationSearchParams.qindex,
+  });
+  const searchInputRef = useRef();
```
Comment (mariiaschv) [inline src/components/AuthoritiesSearchForm/AuthoritiesSearchForm.js:68]:
```suggestion
  });
  
  const searchInputRef = useRef();
```

File: src/components/SearchFilters/SearchFilters.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import { useCallback } from 'react';
+import { useCallback, useContext } from 'react';
```
Comment (mariiaschv) [inline src/components/SearchFilters/SearchFilters.js:1]:
```suggestion
import {
  useCallback,
  useContext,
} from 'react';
```

File: src/components/SearchFilters/SearchFilters.test.js [hunk 3]
```diff
@@ -78,29 +76,30 @@ describe('Given SearchFilters', () => {
 
   describe('when clearing a filter', () => {
     it('should call setFilters with correct filters', () => {
+      mockSetFilters.mockImplementation(setter => setter(defaultCtxValue.filters));
       const { getByText } = renderSearchFilters();
```
Comment (mariiaschv) [inline src/components/SearchFilters/SearchFilters.test.js:80]:
```suggestion
      mockSetFilters.mockImplementation(setter => setter(defaultCtxValue.filters));
      
      const { getByText } = renderSearchFilters();
```

File: src/components/SearchFilters/SearchFilters.test.js [hunk 4]
```diff
@@ -109,6 +108,17 @@ describe('Given SearchFilters', () => {
     });
   });
 
+  describe('when clicking on exclude see from checkbox', () => {
+    it('should call setIsExcludedSeeFromLimiter', () => {
+      mockSetIsExcludedSeeFromLimiter.mockImplementation(setter => setter(defaultCtxValue.isExcludedSeeFromLimiter));
+      const { getByRole } = renderSearchFilters();
```
Comment (mariiaschv) [inline src/components/SearchFilters/SearchFilters.test.js:114]:
```suggestion
      mockSetIsExcludedSeeFromLimiter.mockImplementation(setter => setter(defaultCtxValue.isExcludedSeeFromLimiter));
      
      const { getByRole } = renderSearchFilters();
```

File: src/context/AuthoritiesSearchContext.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+const AuthoritiesSearchContextProvider = ({
+  children,
+}) => {
+  const location = useLocation();
+
+  const nonFilterUrlParams = ['query', 'qindex', 'segment', 'excludeSeeFrom', 'sort'];
+  const getInitialFilters = () => {
+    return omit(buildFiltersObj(location.search), nonFilterUrlParams);
+  };
+  const locationSearchParams = queryString.parse(location.search);
```
Comment (mariiaschv) [inline src/context/AuthoritiesSearchContext.js:36]:
```suggestion
  const nonFilterUrlParams = ['query', 'qindex', 'segment', 'excludeSeeFrom', 'sort'];
  
  const getInitialFilters = () => {
    return omit(buildFiltersObj(location.search), nonFilterUrlParams);
  };
  
  const locationSearchParams = queryString.parse(location.search);
```

File: test/jest/helpers/harness.js [hunk 6]
```diff
@@ -16,13 +20,31 @@ const defaultHistory = createMemoryHistory();
 
 const queryClient = new QueryClient();
 
-const Harness = ({ Router = DefaultRouter, stripes, children, history = defaultHistory }) => {
+const AuthotitiesSearchContextProviderMock = ({ children, ctxValue }) => (
```
Comment (mariiaschv) [inline test/jest/helpers/harness.js:23]: Typo here
```suggestion
const AuthoritiesSearchContextProviderMock = ({ children, ctxValue }) => (
```

**Review by mariiaschv — CHANGES_REQUESTED** [review]
Impressive job! :clap:
Just some small code style changes needed :)

**Review by mariiaschv — APPROVED** [review]

**Review by Svitlana-Korol — APPROVED** [review]

**Pull request merged**